sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala

```diff
@@ -57,7 +57,11 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper {
       // `postScanFilters` and `pushedFilters` can overlap, e.g. the parquet row group filter.
       val (pushedFilters, postScanFiltersWithoutSubquery) = PushDownUtils.pushFilters(
         sHolder.builder, normalizedFiltersWithoutSubquery)
-      val postScanFilters = postScanFiltersWithoutSubquery ++ normalizedFiltersWithSubquery
+      var postScanFilters = postScanFiltersWithoutSubquery ++ normalizedFiltersWithSubquery
+      val partitionFilters = PushDownUtils
+        .pushPartitionFilters(sHolder.builder, sHolder.relation, normalizedFiltersWithoutSubquery)
+      postScanFilters =
+        (ExpressionSet(postScanFilters) -- partitionFilters.filter(_.references.nonEmpty)).toSeq
```

Comment:
Here we have subtracted the `partitionFilters` from the `postScanFilters`. If the file based data source only has partition filters, then the `postScanFilters` should be empty here, and we can push down aggregate for  file based data source.

Reply:
I'm not sure why we can't do this in the old code, where `filters` contain both partition filters and data filters. Suppose we have a method `DataSourceUtils.getPartitionKeyFiltersAndDataFilters`, then we can basically do:

```scala
val (partitionFilters, dataFilters) = DataSourceUtils.getPartitionKeyFiltersAndDataFilters(...)
if (dataFilters.isEmpty) {
  // pushdown aggregates
}
```
?

Reply:
Yes, we can do it that way, but since we have already called `DataSourceUtils.getPartitionKeyFiltersAndDataFilters` to separate the partition filters and data filters, we can just push these two filters to the scan builder, so we can completely remove the v2 partition pruning code from `PruneFileSourcePartitions`.